File: go_engine/board.py
"""
Go board engine for 13x13 (configurable size).

Board representation:
  - 2D numpy array: 0=empty, 1=black, 2=white
  - Zobrist hashing for fast position comparison and ko detection

Coordinate system:
  - (row, col) where (0,0) is top-left
  - "pass" is represented as None
"""

import numpy as np
from collections import deque

EMPTY = 0
BLACK = 1
WHITE = 2

OPPONENT = {BLACK: WHITE, WHITE: BLACK}


def _init_zobrist(size: int) -> np.ndarray:
    """Create Zobrist hash table: random 64-bit ints for each (row, col, color)."""
    rng = np.random.RandomState(42)  # fixed seed for reproducibility
    # shape: (size, size, 3) — index 0 unused, 1=black, 2=white
    return rng.randint(0, 2**63, size=(size, size, 3), dtype=np.uint64)


class Board:
    """
    Go board with full rules: captures, ko, legal move generation, scoring.

    Usage:
        board = Board(size=13)
        board.play(BLACK, (3, 4))
        board.play(WHITE, (2, 3))
        board.play(BLACK, None)  # pass
        legal = board.legal_moves(BLACK)
        score = board.score()
    """

    def __init__(self, size: int = 13):
        self.size = size
        self.grid = np.zeros((size, size), dtype=np.int8)
        self._zobrist_table = _init_zobrist(size)
        self._hash = np.uint64(0)
        # Store full position history for superko detection
        self._position_history: set[int] = set()
        self._position_history.add(int(self._hash))
        self.move_history: list[tuple[int, tuple[int, int] | None]] = []
        self.captured = {BLACK: 0, WHITE: 0}  # stones captured BY each color
        self.komi = 6.5  # compensation for white

    def copy(self) -> "Board":
        """Return a deep copy of this board."""
        new = Board.__new__(Board)
        new.size = self.size
        new.grid = self.grid.copy()
        new._zobrist_table = self._zobrist_table  # shared, immutable
        new._hash = self._hash
        new._position_history = self._position_history.copy()
        new.move_history = self.move_history.copy()
        new.captured = self.captured.copy()
        new.komi = self.komi
        return new

    # ── Neighbors & Groups ──────────────────────────────────────────

    def _neighbors(self, row: int, col: int) -> list[tuple[int, int]]:
        """Return orthogonal neighbors within bounds."""
        result = []
        if row > 0:
            result.append((row - 1, col))
        if row < self.size - 1:
            result.append((row + 1, col))
        if col > 0:
            result.append((row, col - 1))
        if col < self.size - 1:
            result.append((row, col + 1))
        return result

    def _group(self, row: int, col: int) -> tuple[set[tuple[int, int]], set[tuple[int, int]]]:
        """
        BFS to find the connected group containing (row, col) and its liberties.

        Returns:
            (stones, liberties) — sets of coordinates
        """
        color = self.grid[row, col]
        if color == EMPTY:
            return set(), set()

        stones = set()
        liberties = set()
        queue = deque([(row, col)])
        stones.add((row, col))

        while queue:
            r, c = queue.popleft()
            for nr, nc in self._neighbors(r, c):
                if (nr, nc) in stones:
                    continue
                cell = self.grid[nr, nc]
                if cell == EMPTY:
                    liberties.add((nr, nc))
                elif cell == color:
                    stones.add((nr, nc))
                    queue.append((nr, nc))

        return stones, liberties

    def _liberties(self, row: int, col: int) -> int:
        """Count liberties of the group at (row, col)."""
        _, libs = self._group(row, col)
        return len(libs)

    # ── Zobrist Hashing ─────────────────────────────────────────────

    def _toggle_hash(self, row: int, col: int, color: int) -> None:
        """XOR the zobrist value for placing/removing a stone."""
        self._hash ^= self._zobrist_table[row, col, color]

    # ── Stone Placement & Capture ───────────────────────────────────

    def _place_stone(self, color: int, row: int, col: int) -> int:
        """
        Place a stone and capture any opponent groups with zero liberties.

        Returns:
            Number of stones captured.
        """
        self.grid[row, col] = color
        self._toggle_hash(row, col, color)

        captured_count = 0
        opp = OPPONENT[color]

        # Check opponent neighbors for captures
        for nr, nc in self._neighbors(row, col):
            if self.grid[nr, nc] == opp:
                stones, liberties = self._group(nr, nc)
                if len(liberties) == 0:
                    captured_count += len(stones)
                    for sr, sc in stones:
                        self.grid[sr, sc] = EMPTY
                        self._toggle_hash(sr, sc, opp)

        return captured_count

    def _is_suicide(self, color: int, row: int, col: int) -> bool:
        """
        Check if placing color at (row, col) would be suicide.
        A move is suicide if it results in the placed stone's group having zero
        liberties AND it doesn't capture any opponent stones.
        """
        # Temporarily place the stone
        self.grid[row, col] = color

        # Check if any opponent neighbor would be captured
        opp = OPPONENT[color]
        captures = False
        for nr, nc in self._neighbors(row, col):
            if self.grid[nr, nc] == opp:
                _, libs = self._group(nr, nc)
                if len(libs) == 0:
                    captures = True
                    break

        # Check if the placed stone's group has liberties
        _, own_libs = self._group(row, col)

        # Remove the stone
        self.grid[row, col] = EMPTY

        # Suicide if no captures and no liberties
        return not captures and len(own_libs) == 0

    # ── Move Legality ───────────────────────────────────────────────

    def is_legal(self, color: int, move: tuple[int, int] | None) -> bool:
        """
        Check if a move is legal.
        - Pass (None) is always legal.
        - Must be on empty intersection.
        - Must not be suicide.
        - Must not violate positional superko.
        """
        if move is None:
            return True

        row, col = move

        # Bounds check
        if not (0 <= row < self.size and 0 <= col < self.size):
            return False

        # Must be empty
        if self.grid[row, col] != EMPTY:
            return False

        # Suicide check
        if self._is_suicide(color, row, col):
            return False

        # Superko check: simulate the move in-place, check hash, then undo.
        # This avoids a full board copy per intersection.
        self.grid[row, col] = color
        self._toggle_hash(row, col, color)

        # Capture opponent groups with zero liberties
        opp = OPPONENT[color]
        captured_stones: list[tuple[int, int]] = []
        for nr, nc in self._neighbors(row, col):
            if self.grid[nr, nc] == opp:
                stones, liberties = self._group(nr, nc)
                if len(liberties) == 0:
                    for sr, sc in stones:
                        self.grid[sr, sc] = EMPTY
                        self._toggle_hash(sr, sc, opp)
                        captured_stones.append((sr, sc))

        result_hash = int(self._hash)
        is_superko = result_hash in self._position_history

        # Undo: restore captured stones, remove placed stone
        for sr, sc in captured_stones:
            self.grid[sr, sc] = opp
            self._toggle_hash(sr, sc, opp)
        self.grid[row, col] = EMPTY
        self._toggle_hash(row, col, color)

        return not is_superko

    def legal_moves(self, color: int) -> list[tuple[int, int] | None]:
        """
        Return all legal moves for the given color, including pass.
        Pass is always the last element.
        """
        moves = []
        for row in range(self.size):
            for col in range(self.size):
                if self.is_legal(color, (row, col)):
                    moves.append((row, col))
        moves.append(None)  # pass is always legal
        return moves

    # ── Play ────────────────────────────────────────────────────────

    def play(self, color: int, move: tuple[int, int] | None) -> None:
        """
        Play a move. Raises ValueError if illegal.

        Args:
            color: BLACK or WHITE
            move: (row, col) or None for pass
        """
        if not self.is_legal(color, move):
            raise ValueError(f"Illegal move: {move} for {'black' if color == BLACK else 'white'}")

        if move is None:
            self.move_history.append((color, None))
            return

        row, col = move
        captured = self._place_stone(color, row, col)
        self.captured[color] += captured
        self._position_history.add(int(self._hash))
        self.move_history.append((color, (row, col)))

    # ── Scoring (Chinese Rules — Area Scoring) ──────────────────────

    def score(self) -> dict:
        """
        Score the board using Chinese rules (area scoring).
        Territory = empty points surrounded entirely by one color.
        Score = stones on board + territory + komi (for white).

        Returns:
            dict with keys: 'black', 'white', 'winner', 'margin'
        """
        visited = np.zeros((self.size, self.size), dtype=bool)
        territory = {BLACK: 0, WHITE: 0}

        for row in range(self.size):
            for col in range(self.size):
                if visited[row, col] or self.grid[row, col] != EMPTY:
                    continue

                # BFS to find connected empty region
                region = set()
                borders = set()  # colors that border this region
                queue = deque([(row, col)])
                visited[row, col] = True

                while queue:
                    r, c = queue.popleft()
                    region.add((r, c))
                    for nr, nc in self._neighbors(r, c):
                        cell = self.grid[nr, nc]
                        if cell != EMPTY:
                            borders.add(cell)
                        elif not visited[nr, nc]:
                            visited[nr, nc] = True
                            queue.append((nr, nc))

                # Territory belongs to a color only if bordered by exactly one color
                if len(borders) == 1:
                    owner = borders.pop()
                    territory[owner] += len(region)

        # Count stones on board
        black_stones = int(np.sum(self.grid == BLACK))
        white_stones = int(np.sum(self.grid == WHITE))

        black_score = black_stones + territory[BLACK]
        white_score = white_stones + territory[WHITE] + self.komi

        if black_score > white_score:
            winner = "black"
        elif white_score > black_score:
            winner = "white"
        else:
            winner = "draw"

        return {
            "black": black_score,
            "white": white_score,
            "winner": winner,
            "margin": abs(black_score - white_score),
        }

    # ── Display ─────────────────────────────────────────────────────

    def __str__(self) -> str:
        """Pretty-print the board."""
        symbols = {EMPTY: ".", BLACK: "X", WHITE: "O"}
        col_labels = "  " + " ".join(
            chr(ord("A") + i + (1 if i >= 8 else 0)) for i in range(self.size)
        )
        rows = [col_labels]
        for r in range(self.size):
            row_label = f"{self.size - r:2d}"
            row_str = " ".join(symbols[self.grid[r, c]] for c in range(self.size))
            rows.append(f"{row_label} {row_str} {row_label}")
        rows.append(col_labels)
        return "\n".join(rows)

File: src/board.cpp
#include "board.h"
#include <cstring>
#include <random>
#include <queue>
#include <algorithm>

namespace alphago {

Board::Board(int size, bool with_history) : size_(size), hash_(0), move_count_(0) {
    std::memset(grid_, 0, sizeof(grid_));
    captured_[0] = 0;
    captured_[1] = 0;
    captured_[2] = 0;
    track_history = with_history;
    init_zobrist();
    if (with_history) {
        position_history_.insert(hash_);
    }
}

Board::Board(const Board& other)
    : size_(other.size_)
    , hash_(other.hash_)
    , position_history_(other.position_history_)
    , move_count_(other.move_count_)
    , komi(other.komi)
{
    std::memcpy(grid_, other.grid_, sizeof(grid_));
    std::memcpy(zobrist_table_, other.zobrist_table_, sizeof(zobrist_table_));
    captured_[0] = other.captured_[0];
    captured_[1] = other.captured_[1];
    captured_[2] = other.captured_[2];
}

Board& Board::operator=(const Board& other) {
    if (this != &other) {
        size_ = other.size_;
        hash_ = other.hash_;
        position_history_ = other.position_history_;
        move_count_ = other.move_count_;
        komi = other.komi;
        std::memcpy(grid_, other.grid_, sizeof(grid_));
        std::memcpy(zobrist_table_, other.zobrist_table_, sizeof(zobrist_table_));
        captured_[0] = other.captured_[0];
        captured_[1] = other.captured_[1];
        captured_[2] = other.captured_[2];
    }
    return *this;
}

Board Board::copy_light() const {
    // Create with history disabled — position_history_ stays empty (zero heap alloc).
    Board b(size_, /*with_history=*/false);
    std::memcpy(b.grid_, grid_, sizeof(grid_));
    std::memcpy(b.zobrist_table_, zobrist_table_, sizeof(zobrist_table_));
    b.hash_ = hash_;
    b.captured_[0] = captured_[0];
    b.captured_[1] = captured_[1];
    b.captured_[2] = captured_[2];
    b.move_count_ = move_count_;
    b.komi = komi;
    return b;
}

void Board::init_zobrist() {
    // Fixed seed for reproducibility (matches Python's RandomState(42))
    std::mt19937_64 rng(42);
    for (int r = 0; r < MAX_BOARD_SIZE; r++) {
        for (int c = 0; c < MAX_BOARD_SIZE; c++) {
            zobrist_table_[r][c][0] = 0;
            zobrist_table_[r][c][1] = rng();
            zobrist_table_[r][c][2] = rng();
        }
    }
}

void Board::toggle_hash(int r, int c, uint8_t color) {
    hash_ ^= zobrist_table_[r][c][color];
}

Board::Neighbors Board::neighbors(int r, int c) const {
    Neighbors n;
    n.count = 0;
    if (r > 0)          n.data[n.count++] = {r - 1, c};
    if (r < size_ - 1)  n.data[n.count++] = {r + 1, c};
    if (c > 0)          n.data[n.count++] = {r, c - 1};
    if (c < size_ - 1)  n.data[n.count++] = {r, c + 1};
    return n;
}

Board::GroupInfo Board::find_group(int r, int c) const {
    GroupInfo info;
    uint8_t color = grid_[r][c];
    if (color == EMPTY) return info;

    bool visited[MAX_BOARD_SIZE][MAX_BOARD_SIZE] = {};
    bool lib_visited[MAX_BOARD_SIZE][MAX_BOARD_SIZE] = {};

    std::queue<std::pair<int,int>> queue;
    queue.push({r, c});
    visited[r][c] = true;
    info.stones.push_back({r, c});

    while (!queue.empty()) {
        auto [cr, cc] = queue.front();
        queue.pop();

        auto nbrs = neighbors(cr, cc);
        for (int i = 0; i < nbrs.count; i++) {
            auto [nr, nc] = nbrs.data[i];
            if (grid_[nr][nc] == EMPTY && !lib_visited[nr][nc]) {
                lib_visited[nr][nc] = true;
                info.liberties.push_back({nr, nc});
            } else if (grid_[nr][nc] == color && !visited[nr][nc]) {
                visited[nr][nc] = true;
                info.stones.push_back({nr, nc});
                queue.push({nr, nc});
            }
        }
    }

    return info;
}

int Board::count_liberties(int r, int c) const {
    return static_cast<int>(find_group(r, c).liberties.size());
}

bool Board::has_liberties(int r, int c, int min_libs) const {
    // BFS using stack-allocated arrays — zero heap allocation.
    uint8_t color = grid_[r][c];
    if (color == EMPTY) return true;

    bool visited[MAX_BOARD_SIZE][MAX_BOARD_SIZE] = {};
    bool lib_visited[MAX_BOARD_SIZE][MAX_BOARD_SIZE] = {};

    // Stack-allocated queue (max group size = board area)
    std::pair<int,int> queue_buf[MAX_BOARD_SIZE * MAX_BOARD_SIZE];
    int queue_head = 0, queue_tail = 0;

    visited[r][c] = true;
    queue_buf[queue_tail++] = {r, c};
    int lib_count = 0;

    while (queue_head < queue_tail) {
        auto [cr, cc] = queue_buf[queue_head++];
        auto nbrs = neighbors(cr, cc);
        for (int i = 0; i < nbrs.count; i++) {
            auto [nr, nc] = nbrs.data[i];
            if (grid_[nr][nc] == EMPTY && !lib_visited[nr][nc]) {
                lib_visited[nr][nc] = true;
                lib_count++;
                if (lib_count >= min_libs) return true;
            } else if (grid_[nr][nc] == color && !visited[nr][nc]) {
                visited[nr][nc] = true;
                queue_buf[queue_tail++] = {nr, nc};
            }
        }
    }
    return lib_count >= min_libs;
}

int Board::place_stone(uint8_t color, int r, int c) {
    grid_[r][c] = color;
    toggle_hash(r, c, color);

    int captured_count = 0;
    uint8_t opp = opponent(color);

    auto nbrs = neighbors(r, c);
    for (int i = 0; i < nbrs.count; i++) {
        auto [nr, nc] = nbrs.data[i];
        if (grid_[nr][nc] == opp) {
            // First check if the group has any liberties (stack-allocated, fast)
            if (!has_liberties(nr, nc, 1)) {
                // Group is captured — find stones to remove (need full find_group)
                auto group = find_group(nr, nc);
                captured_count += static_cast<int>(group.stones.size());
                for (auto [sr, sc] : group.stones) {
                    grid_[sr][sc] = EMPTY;
                    toggle_hash(sr, sc, opp);
                }
            }
        }
    }

    return captured_count;
}

bool Board::is_suicide(uint8_t color, int r, int c) const {
    // Temporarily place stone (const_cast for the check, then undo)
    auto* self = const_cast<Board*>(this);
    self->grid_[r][c] = color;

    uint8_t opp = opponent(color);
    bool captures = false;
    auto nbrs = neighbors(r, c);
    for (int i = 0; i < nbrs.count; i++) {
        auto [nr, nc] = nbrs.data[i];
        if (grid_[nr][nc] == opp) {
            // Use has_liberties (stack-allocated BFS, no heap alloc)
            if (!has_liberties(nr, nc, 1)) {
                captures = true;
                break;
            }
        }
    }

    // Check own liberties (stack-allocated BFS)
    bool own_has_libs = has_liberties(r, c, 1);

    self->grid_[r][c] = EMPTY;

    return !captures && !own_has_libs;
}

bool Board::is_legal(uint8_t color, Move move) const {
    if (move.is_pass()) return true;

    int r = move.row, c = move.col;
    if (r < 0 || r >= size_ || c < 0 || c >= size_) return false;
    if (grid_[r][c] != EMPTY) return false;

    // Fast path: if any adjacent cell is empty, it's not suicide
    // and we only need the superko check
    bool has_adjacent_liberty = false;
    auto nbrs = neighbors(r, c);
    for (int i = 0; i < nbrs.count; i++) {
        if (grid_[nbrs.data[i].first][nbrs.data[i].second] == EMPTY) {
            has_adjacent_liberty = true;
            break;
        }
    }

    if (!has_adjacent_liberty) {
        // Need full suicide check
        if (is_suicide(color, r, c)) return false;
    }

    // Skip superko check for simulation boards (no history tracking).
    // This avoids the expensive simulate-check-undo cycle AND
    // the vector<pair> heap allocation for captured_stones.
    if (!track_history) {
        return true;  // passed bounds, occupied, and suicide checks
    }

    // Superko check: simulate in-place, check hash, undo
    auto* self = const_cast<Board*>(this);
    self->grid_[r][c] = color;
    self->toggle_hash(r, c, color);

    uint8_t opp = opponent(color);
    std::vector<std::pair<int,int>> captured_stones;

    for (int i = 0; i < nbrs.count; i++) {
        auto [nr, nc] = nbrs.data[i];
        if (self->grid_[nr][nc] == opp) {
            auto group = find_group(nr, nc);
            if (group.liberties.empty()) {
                for (auto [sr, sc] : group.stones) {
                    self->grid_[sr][sc] = EMPTY;
                    self->toggle_hash(sr, sc, opp);
                    captured_stones.push_back({sr, sc});
                }
            }
        }
    }

    uint64_t result_hash = self->hash_;
    bool is_superko = position_history_.count(result_hash) > 0;

    // Undo
    for (auto [sr, sc] : captured_stones) {
        self->grid_[sr][sc] = opp;
        self->toggle_hash(sr, sc, opp);
    }
    self->grid_[r][c] = EMPTY;
    self->toggle_hash(r, c, color);

    return !is_superko;
}

void Board::play(uint8_t color, Move move) {
    if (move.is_pass()) {
        move_count_++;
        return;
    }

    int captured = place_stone(color, move.row, move.col);
    captured_[color] += captured;
    if (track_history) {
        position_history_.insert(hash_);
    }
    move_count_++;
}

std::vector<Move> Board::legal_moves(uint8_t color) const {
    std::vector<Move> moves;
    moves.reserve(size_ * size_ + 1);

    for (int r = 0; r < size_; r++) {
        for (int c = 0; c < size_; c++) {
            if (is_legal(color, board_move(r, c))) {
                moves.push_back(board_move(r, c));
            }
        }
    }
    moves.push_back(pass_move()); // pass always legal
    return moves;
}

Board::Score Board::score() const {
    bool visited[MAX_BOARD_SIZE][MAX_BOARD_SIZE] = {};
    int territory[3] = {};

    for (int r = 0; r < size_; r++) {
        for (int c = 0; c < size_; c++) {
            if (visited[r][c] || grid_[r][c] != EMPTY) continue;

            // BFS to find empty region
            std::vector<std::pair<int,int>> region;
            std::unordered_set<uint8_t> borders;
            std::queue<std::pair<int,int>> queue;
            queue.push({r, c});
            visited[r][c] = true;

            while (!queue.empty()) {
                auto [cr, cc] = queue.front();
                queue.pop();
                region.push_back({cr, cc});

                auto nbrs = neighbors(cr, cc);
                for (int i = 0; i < nbrs.count; i++) {
                    auto [nr, nc] = nbrs.data[i];
                    uint8_t cell = grid_[nr][nc];
                    if (cell != EMPTY) {
                        borders.insert(cell);
                    } else if (!visited[nr][nc]) {
                        visited[nr][nc] = true;
                        queue.push({nr, nc});
                    }
                }
            }

            if (borders.size() == 1) {
                uint8_t owner = *borders.begin();
                territory[owner] += static_cast<int>(region.size());
            }
        }
    }

    int black_stones = 0, white_stones = 0;
    for (int r = 0; r < size_; r++)
        for (int c = 0; c < size_; c++) {
            if (grid_[r][c] == BLACK) black_stones++;
            else if (grid_[r][c] == WHITE) white_stones++;
        }

    Score s;
    s.black_stones = black_stones;
    s.white_stones = white_stones;
    s.black_territory = territory[BLACK];
    s.white_territory = territory[WHITE];
    s.black = black_stones + territory[BLACK];
    s.white = white_stones + territory[WHITE] + komi;
    s.margin = std::abs(s.black - s.white);
    if (s.black > s.white) s.winner = BLACK;
    else if (s.white > s.black) s.winner = WHITE;
    else s.winner = 0;

    return s;
}

} // namespace alphago
